File: quakepulse/data/parser.py
"""Parse USGS GeoJSON features into structured tabular data."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

import pandas as pd


def earthquakes_to_df(eq_list: list[dict[str, Any]]) -> pd.DataFrame:
    """Convert USGS GeoJSON feature list to a normalized DataFrame."""
    records: list[dict[str, Any]] = []

    for eq in eq_list:
        props = eq.get("properties") or {}
        geometry = eq.get("geometry") or {}
        coordinates = geometry.get("coordinates") or [None, None, None]

        mag = props.get("mag")
        magnitude = float(mag) if mag is not None else 0.0

        time_ms = props.get("time")
        if time_ms is None:
            continue

        records.append(
            {
                "ID": eq.get("id", ""),
                "Magnitude": magnitude,
                "Depth": float(coordinates[2]) if coordinates[2] is not None else 0.0,
                "Place": props.get("place") or "Unknown location",
                "Time": datetime.fromtimestamp(time_ms / 1000, tz=timezone.utc),
                "Latitude": float(coordinates[1]) if coordinates[1] is not None else 0.0,
                "Longitude": float(coordinates[0]) if coordinates[0] is not None else 0.0,
                "Status": props.get("status") or "unknown",
                "Type": props.get("type") or "earthquake",
                "Url": props.get("url") or "",
            }
        )

    if not records:
        return pd.DataFrame(
            columns=[
                "ID",
                "Magnitude",
                "Depth",
                "Place",
                "Time",
                "Latitude",
                "Longitude",
                "Status",
                "Type",
                "Url",
            ]
        )

    return pd.DataFrame(records)

File: server/src/services/parser.js
/**
 * Normalize USGS GeoJSON features into structured event records.
 */

export function parseFeatures(features = []) {
  if (!Array.isArray(features)) return [];
  const events = [];

  for (const feature of features) {
    const props = feature?.properties || {};
    const coords = feature?.geometry?.coordinates;
    const timeMs = props.time;

    if (timeMs == null || !Number.isFinite(Number(timeMs))) continue;
    if (!Array.isArray(coords) || coords.length < 2) continue;

    const magRaw = props.mag == null ? 0 : Number(props.mag);
    const lon = Number(coords[0]);
    const lat = Number(coords[1]);
    const depthRaw = coords[2] == null ? 0 : Number(coords[2]);

    if (!Number.isFinite(lat) || !Number.isFinite(lon)) continue;
    if (lat < -90 || lat > 90 || lon < -180 || lon > 180) continue;

    const magnitude = Number.isFinite(magRaw) ? magRaw : 0;
    const depth = Number.isFinite(depthRaw) ? depthRaw : 0;

    events.push({
      id: feature.id || `${lat}-${lon}-${timeMs}`,
      magnitude,
      depth,
      place: props.place || 'Unknown location',
      time: new Date(timeMs).toISOString(),
      timeMs: Number(timeMs),
      latitude: lat,
      longitude: lon,
      status: props.status || 'unknown',
      type: props.type || 'earthquake',
      url: props.url || '',
      felt: props.felt ?? null,
      cdi: props.cdi ?? null,
      mmi: props.mmi ?? null,
      tsunami: Boolean(props.tsunami),
      sig: props.sig ?? null,
      alert: props.alert || null,
    });
  }

  return events;
}

export function enrichEvents(events) {
  return events.map((e) => {
    const magnitude = e.magnitude;
    const energyMag = Math.max(0, magnitude);
    const energyJ = 10 ** (1.5 * energyMag + 4.8);
    const magRange = magnitude < 2.5 ? 'Low' : magnitude < 5 ? 'Medium' : 'High';
    const magCategory =
      magnitude < 3 ? 'Minor' : magnitude < 6 ? 'Moderate' : magnitude < 9 ? 'Strong' : 'Severe';
    const depthCategory =
      e.depth < 70 ? 'Shallow' : e.depth < 300 ? 'Intermediate' : 'Deep';

    return {
      ...e,
      sizeScaled: (Math.max(0, magnitude) + 0.3) * 5,
      energyJ,
      magRange,
      magCategory,
      depthCategory,
      region: extractRegion(e.place),
    };
  });
}

export function extractRegion(place = '') {
  const text = String(place).trim();
  if (!text) return 'Unknown';

  const parts = text
    .split(',')
    .map((p) => p.trim())
    .filter(Boolean);
  if (parts.length > 1) return parts[parts.length - 1];

  // "Off coast of Japan", "Near coast of Chile", "10 km N of Tokyo"
  const ofMatch = text.match(/\bof\s+(.+)$/i);
  if (ofMatch?.[1]) return ofMatch[1].trim();

  return parts[0] || text;
}
